prov/efa/src/rdm/efa_rdm_pke_utils.h

```diff
@@ -94,6 +94,80 @@ efa_rdm_pke_copy_from_hmem_iov(struct efa_mr *iov_mr, struct efa_rdm_pke *pke,
 	return copied;
 }
 
+/** 
+ * @brief This function either posts RDMA read, or sends a NACK packet when p2p
+ * is not available or memory registration limit was reached on the receiver.
+ *
+ * @param[in]    ep         endpoint
+ * @param[in]    pkt_entry  packet entry
+ * @param[in]    rxe        RX entry
+ *
+ * @return 0 on success, or a negative error code.
+ */
+static inline int
+efa_rdm_pke_post_remote_read_or_nack(struct efa_rdm_ep *ep,
+				     struct efa_rdm_pke *pkt_entry,
+				     struct efa_rdm_ope *rxe)
+{
+	int err = 0;
+	int pkt_type;
+	int p2p_avail;
+
+	pkt_type = efa_rdm_pke_get_base_hdr(pkt_entry)->type;
+	err = efa_rdm_ep_use_p2p(ep, rxe->desc[0]);
+	if (err < 0)
+		return err;
+
+	p2p_avail = err;
+	if (p2p_avail) {
+		err = efa_rdm_ope_post_remote_read_or_queue(rxe);
+	} else if (efa_rdm_peer_support_read_nack(rxe->peer)) {
+		EFA_INFO(FI_LOG_EP_CTRL,
+			 "Receiver sending long read "
+			 "NACK packet because P2P is not available, "
+			 "unable to post RDMA read.\n");
+		goto send_nack;
+	} else {
+		EFA_INFO(FI_LOG_EP_CTRL, "P2P is not available, "
+					 "unable to post RDMA read.\n");
+		err = -FI_EOPNOTSUPP;
+	}
+
+	if (err == -FI_ENOMR) {
+		if (efa_rdm_peer_support_read_nack(rxe->peer)) {
+			EFA_INFO(FI_LOG_EP_CTRL, "Receiver sending long read "
+						 "NACK packet because memory "
+						 "registration limit was "
+						 "reached on the receiver.\n");
+			goto send_nack;
+		} else {
+			/* Peer does not support the READ_NACK packet. So we
+			 * return EAGAIN and hope that the app runs progress
+			 * again which will free some MR registrations */
+			err = -FI_EAGAIN;
+		}
+	}
+
+	return err;
```

Comment:
similar comments, I think you can return the error code in the earlier blocks directly?